### ISSUE TITLE
Write jax instructions

### ISSUE THREAD

**Issue by andrewni2002** [andrewni420/GPU-PUSH#1]
- [ ] Devise a reasonable set of jax functions (maybe all jax.numpy/jax.lax functions?) to wrap as instructions for potential use in push.
    - [ ] Do we want variants of functions, e.g. {log} vs {log, log10, log2, log1p}?

Implement instruction signatures:
- These are functions to calculate the shape of the output array given the shape(s) of the input array(s)
- [x] Broadcasting
- [x] Matrix multiplication
- [x] Convolution
    - [x] Pooling
- [x] Aggregating
- [x] Casting
- [ ] Reshaping
    - [x] Flattening 
    - [x] Chunking
- [ ] Reordering
    - [x] Swap axes

Implement instructions (Entries with a question mark are up for debate):
- [ ] Decide on whether we even want to include the possibility of eager execution
- "Zero"-ary operations
    - [ ] arange, zeros, ones, full, eye
    - [ ] tri, tril tril_indices, triu, triu_indices (?)
- Unary operations
    - Standard functions
        - [x] sin, cos, log, exp, abs, negative, reciprocal
        - [ ] square, cube, sqrt cbrt, sign, sinc (?)
        - [ ] sinh, cosh, tanh, csc, sec (?)
        - [ ] arccos, arcsin, arctan
    - Logic functions (?)
        - [ ] logical_not 
    - Casting
        - [x] float to int, int to float
    - Indexing
        - [ ] argmax, argmin, argsort, argpartition, argwhere
        - [ ] partition, sort, 
        - [ ] scatter, gather, select, take, take_along_axis (?)
        - [ ] scatter_add, scatter_max, scatter_min, scatter_mul (?)
        - [ ] extract, compress (?)
        - [ ] indices (?)
        - [ ] mask_indices (?)
    - Aggregating
        - [ ] sum, average, max, min, var, std
        - [ ] prod (?)
        - [ ] cumsum, cumprod, cummax, cummin, diff (?)
        - [ ] top_k, approx_max_k, approx_min_k (?)
    - Reshaping / reordering / misc
        - [x] chunk
        - [ ] swapaxes, transpose, moveaxis
        - [ ] concatenate (?)
        - [ ] copy
        - [ ] expand_dims, squeeze (?)
        - [ ] reshape
            - [x] flatten
        - [ ] flip (?)
        - [ ] pad
        - [ ] roll (?)
        - [ ] rot90 (?)
        - [ ] stop_gradient (?)
    - Activation functions
        - [x] relu, relu6, sigmoid, softplus, sparse_plus, etc. from jax.nn
- Binary operations 
    - Standard functions
        - [x] add, subtract, multiply, divide, matrix multiply, inner/outer product, convolution
        - [x] maximum, minimum
        - [ ] logaddexp, logsumexp
        - [ ] power (?)
        - [ ] tensordot (?)
    - Logic functions (?)
        - [ ] logical_and, logical_or, logical_xor
        - [ ] greater, greater_equal, less, less_equal, isclose
- Ternary operations
    - [ ] where, cond 
    - [ ] 

Write tests for instructions:
- [ ] Refactor tests to read test cases from an input file
- [ ] Refactor tests to make it easier to test new instructions
- Test each instruction:
